instrument/invariant.go

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package instrument
+
+const (
+	// InvariantViolatedMetricName is the name of the metric emitted upon
+	// invocation of `EmitInvariantViolation`.
+	InvariantViolatedMetricName = "invariant-violated"
+
+	// InvariantViolatedLogPrefix is a common prefix that should be used
+	// when generating errors/log statements pertaining to the violation
+	// of an invariant.
+	InvariantViolatedLogPrefix = "[invariant violated]"
+)
+
+// EmitInvariantViolation emits a metric to indicate a system invariant has
+// been violated. Users of this method are expected to monitor/alert off this
+// metric to ensure they're notified when such an event occurs. Further, they
+// should log further information to aid diagnostics of the system invariant
+// violated at the callsite of the violation.
+func EmitInvariantViolation(opts Options) {
+	// NB(prateek): there's no need to cache this metric. It should be never
```

Comment:
I don't have much context here, but would it be preferable to just crash in the event of an invariant being violated and rely on the underlying process manager to restart the process rather than try to recover when certain assumptions may no longer be valid?

Reply:
The motivation for this P.R came up in response to the data loss investigation I’ve been working on. As a result of that, we found several places in the codebase where we could add additional safeguards to detect issues.

A good number of these cases are bad enough to cause data loss, but if they are ever triggered they represent fundamental issues with M3DB that need to be addressed with bug fixes, and there would be no value in causing additional data loss by crashing the node. When stuff like this happens we emit logs, but we have no idea if these logs are being emitted on a regular basis unless we notice something strange and go look at them. With this metric we can have a low-priority alert so that we’re at least notified that the issue is actually manifesting in production.

In other situations where we’ve clearly reached a state that we should not be in and cannot proceed I agree we should just crash and restart.

Reply:
Pretty much what Richie said. 

The choice of when to run in a degraded state v crashing needs to be made before using the utilities in this PR. This PR is to make a standard way to detect when we’ve chosen to run in a degraded state.

Reply:
Gotcha, I think my hangup was  just with the word invariant. The connotation I have when I hear invariant is that the property can never be violated. Any thoughts on calling it an assumption or something similar? Don’t feel too strongly here though

Reply:
That's definitely intentional. This is going to used in places where we're doing the equivalent of an assert and can still continue processing (e.g. return error response to the user for a RPC code path that runs into an un-expected path). Going to retain the name to indicate severity.